phoenix-tools/src/main/java/org/apache/phoenix/schema/SchemaExtractionProcessor.java

```diff
@@ -0,0 +1,410 @@
+        for (PColumn indexedColumn : indexPK) {
+            String indexColumn = extractIndexColumn(indexedColumn.getName().getString(), defaultCF);
+            if(indexColumn.equalsIgnoreCase(MetaDataUtil.VIEW_INDEX_ID_COLUMN_NAME)) {
+                continue;
+            }
+            indexPkSet.add(indexColumn);
+            sortOrderMap.put(indexColumn, indexedColumn.getSortOrder());
+        }
+
+        for(PColumn pColumn : dataPK) {
+            dataPkSet.add(pColumn.getName().getString());
+        }
+
+        Set<String> effectivePK = Sets.symmetricDifference(indexPkSet, dataPkSet);
+        if (effectivePK.isEmpty()) {
+            effectivePK = indexPkSet;
+        }
+        for (String column : effectivePK) {
+            if(indexedColumnsBuilder.length()!=0) {
+                indexedColumnsBuilder.append(", ");
+            }
+            indexedColumnsBuilder.append(column);
+            if(sortOrderMap.get(column)!= SortOrder.getDefault()) {
+                indexedColumnsBuilder.append(" ");
+                indexedColumnsBuilder.append(sortOrderMap.get(column));
+            }
+        }
+        return indexedColumnsBuilder.toString();
+    }
+
+    private String extractIndexColumn(String columnName, String defaultCF) {
+        String [] columnNameSplit = columnName.split(":");
+        if(columnNameSplit[0].equals("") || columnNameSplit[0].equalsIgnoreCase(defaultCF)) {
+            return columnNameSplit[1];
+        } else {
+            return columnName.replace(":", ".");
+        }
+    }
+
+    private String getCoveredColumnsString(PTable indexPTable, String defaultCF) {
+        StringBuilder coveredColumnsBuilder = new StringBuilder();
+        List<PColumn> pkColumns = indexPTable.getColumns();
+        for (PColumn cc : pkColumns) {
+            if(coveredColumnsBuilder.length()!=0) {
+                coveredColumnsBuilder.append(", ");
+            }
+            if(cc.getFamilyName()!=null) {
+                String indexColumn = extractIndexColumn(cc.getName().getString(), defaultCF);
+                coveredColumnsBuilder.append(indexColumn);
+            }
+        }
+        return coveredColumnsBuilder.toString();
+    }
+
+    protected String generateIndexDDLString(String baseTableFullName, String indexedColumnString,
+            String coveredColumnString, boolean local, String pTableName) {
+        StringBuilder outputBuilder = new StringBuilder(String.format(CREATE_INDEX,
+                local ? "LOCAL " : "", pTableName, baseTableFullName));
+        outputBuilder.append("(");
+        outputBuilder.append(indexedColumnString);
+        outputBuilder.append(")");
+        if(!coveredColumnString.equals("")) {
+            outputBuilder.append(" INCLUDE (");
+            outputBuilder.append(coveredColumnString);
+            outputBuilder.append(")");
+        }
+        return outputBuilder.toString();
+    }
+
+    PTable getPTable(String pTableFullName) throws SQLException {
```

Comment:
This won't work for tenant-owned schema objects -- if you're not going to tackle this now please file a subtask JIRA to do so before we merge into a real branch

Reply:
Would like to get it fixed in subsequent commits. Yes, Let me file one.

Reply:
I fixed it in the same commit.